### ISSUE TITLE
Dataset versioning

### ISSUE THREAD

**Issue by gplssm** [openego/eGon-data#75]
In a longer call @gnn and @gplssm discussed how to introduce dataset versions, how to build and store the relationship between dataset versions and a related task, and which use cases are covered by introducing data versions.

_proposals which we have agreed on during the call. Italic text is an addition which needs to be discussed_

- The package version of eGon-data (respectively the commit ID) defines the database table structure
- Datasets (single tables or groups of tables) get a version tag. For example we introduce a column "version" for each database table
- At the beginning of each task if the expected version of data does exist
- _If a dataset consists of several tables and only one version is given, each of the tables is checked for the existence of the desired version_
- Options for store information about the relationship of dataset version to a task
  - Airflow lineage
  - `.yml` file (may integrated into `datasets.yml`)

We need to discuss further on that. For now, please add aspects I forgot and also add more questions if you have some @gnn 

### Open questions

- Should versioning be done per dataset or per table (of a dataset)?

**Comment by ulfmueller** [comment]
It seems to me that the following sentence is missing s.th.:

> At the beginning of each task if the expected version of data does exist

Do you mean that at the beginning of each task it is checked whether the desired version data already exists? If it is exists it does not re-execute the process which would overwrite the data but instead uses the existing data?

**Comment by ulfmueller** [comment]
additionally some hopefully helpful thoughts:

I think it is important to specify the different 'layers' - the local one and the OEP one. Locally it can/should be checked for existing versioned and more over it could be checked for existing versions on the OEP. It could be an option that one cannot create data with the same version id already existing in the OEP - then it would also not be possible to double up / overwrite data on the OEP. If instead the local versioning is not directly coupled with the one on the OEP, I think it is important to assure that once the local process wants to load data to the OEP it can only do that for data with a new version number.

I also wonder if it could be a problem, that if two or more local users are trying to load versioned data to the OEP basically at the same time. It should be prevented that they load up data with the same versions resulting in a crazy mixture of data.

**Comment by gplssm** [comment]
> It seems to me that the following sentence is missing s.th.:
> 
> > At the beginning of each task if the expected version of data does exist
> 
> Do you mean that at the beginning of each task it is checked whether the desired version data already exists? If it is exists it does not re-execute the process which would overwrite the data but instead uses the existing data?

Yes

**Comment by ulfmueller** [comment]
Is there already a .yml file or s.th. where a first dataset version is defined? I would like to fill some grid.xyz tables and I defined that the version column may not be null. Therefore I need to specify a version tag. Is there already a 'beta-workflow' for that? If not, do you have any ideas/propositions?

**Comment by gplssm** [comment]
Nope, no solution ready yet. You could specify a field `version` in `datasets.yml` and check against that/ fill table with the version specified there.
